internal/reconciliations/gateway/kyma_gateway_test.go

```diff
@@ -199,151 +226,123 @@ var _ = Describe("Kyma Gateway reconciliation", func() {
-			})
+			}, true, controllers.Ready, BeTrue(), Not(ContainElement(KymaGatewayFinalizer)))
 		})
 
-	})
-
-	It("should not delete Kyma Gateway when EnableKymaGateway is updated to false, but any APIRule exists", func() {
-		// given
-		apiGateway := getApiGateway(true)
-
-		apiRule := v1beta1.APIRule{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: "test",
-			},
-		}
-
-		k8sClient := createFakeClient(&apiGateway, &apiRule)
-		status := ReconcileKymaGateway(context.Background(), k8sClient, &apiGateway)
-		Expect(status.IsReady()).To(BeTrue())
-		kymaGateway := v1alpha3.Gateway{}
-		Expect(k8sClient.Get(context.Background(), client.ObjectKey{Name: kymaGatewayName, Namespace: kymaGatewayNamespace}, &kymaGateway)).Should(Succeed())
+		It("Should not delete Kyma Gateway, DNSEntry and Certificate and finalizer when EnableKymaGateway is updated to false but there is blocking APIRule", func() {
+			apiRule := getApiRule(KymaGatewayFullName)
+			status := testShouldDeleteKymaGatewayResources(func(gw v1alpha1.APIGateway) v1alpha1.APIGateway {
+				gw.Spec.EnableKymaGateway = ptr.To(false)
+				return gw
+			}, true, controllers.Warning, BeFalse(), ContainElement(KymaGatewayFinalizer), &apiRule)
 
-		updatedApiGateway := v1alpha1.APIGateway{
-			ObjectMeta: metav1.ObjectMeta{
-				Name: "test",
-			},
-		}
+			Expect(status.NestedError().Error()).To(Equal("could not delete Kyma Gateway since there are 1 custom resource(s) present that block its deletion"))
+			Expect(status.Description()).To(Equal("There are custom resources that block the deletion. Please take a look at kyma-system/api-gateway-controller-manager logs to see more information about the warning"))
+		})
 
-		// when
-		status = ReconcileKymaGateway(context.Background(), k8sClient, &updatedApiGateway)
+		It("Should not delete Kyma Gateway, DNSEntry and Certificate and finalizer when EnableKymaGateway is updated to false but there is blocking VirtualService", func() {
+			vs := getVirtualService(KymaGatewayFullName)
+			status := testShouldDeleteKymaGatewayResources(func(gw v1alpha1.APIGateway) v1alpha1.APIGateway {
+				gw.Spec.EnableKymaGateway = ptr.To(false)
+				return gw
+			}, true, controllers.Warning, BeFalse(), ContainElement(KymaGatewayFinalizer), &vs)
 
-		// then
-		Expect(status.IsWarning()).To(BeTrue())
-		err := k8sClient.Get(context.Background(), client.ObjectKey{Name: kymaGatewayName, Namespace: kymaGatewayNamespace}, &kymaGateway)
-		Expect(err).ShouldNot(HaveOccurred())
+			Expect(status.NestedError().Error()).To(Equal("could not delete Kyma Gateway since there are 1 custom resource(s) present that block its deletion"))
+			Expect(status.Description()).To(Equal("There are custom resources that block the deletion. Please take a look at kyma-system/api-gateway-controller-manager logs to see more information about the warning"))
+		})
 	})
-
 })
 
-func testShouldDeleteKymaGatewayNonGardenerResources(updateApiGateway func(gw v1alpha1.APIGateway) v1alpha1.APIGateway) {
```

Comment:
Yes, I agree but in this case it solves quite a code duplication and adding more cases avoids also adding more code duplication. Let me check if it could be better with less boolean-controlled logic and less code duplication.

Reply:
I always consider readability more important than code duplication (especially in tests), since there will always be overlapping assertion scenarios. I have seen tests that called multiple functions for asserting resources. These tests had no code duplication, but it was a mess to understand all assertions are done for one specific test.

If it's too much effort, we don't have to do it now.